File: vane-core/src/main/java/org/oddlama/vane/core/lang/LangField.java
package org.oddlama.vane.core.lang;

import java.lang.reflect.Field;
import java.util.function.Function;
import org.bukkit.configuration.file.YamlConfiguration;
import org.oddlama.vane.core.YamlLoadException;
import org.oddlama.vane.core.module.Module;
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator;

public abstract class LangField<T> {

    public static final String PREFIX = "lang";

    private Module<?> module;
    protected Object owner;
    protected Field field;
    protected String name;
    private final String namespace;
    private final String key;

    public LangField(Module<?> module, Object owner, Field field, Function<String, String> mapName) {
        this.module = module;
        this.owner = owner;
        this.field = field;

        if (!field.getName().startsWith(PREFIX)) throw new RuntimeException(
            new YamlLoadException.Lang("field must start with " + PREFIX, this)
        );
        this.name = mapName.apply(field.getName().substring(PREFIX.length()));
        this.namespace = module.namespace();
        this.key = namespace + "." + yamlPath();

        field.setAccessible(true);
    }

    public String getName() {
        return name;
    }

    public String yamlPath() {
        return name;
    }

    protected void checkYamlPath(YamlConfiguration yaml) throws YamlLoadException {
        if (!yaml.contains(name, true)) {
            throw new YamlLoadException.Lang("yaml is missing entry with path '" + name + "'", this);
        }
    }

    public Module<?> module() {
        return module;
    }

    public String namespace() {
        return namespace;
    }

    public String key() {
        return key;
    }

    public abstract void checkLoadable(YamlConfiguration yaml) throws YamlLoadException;

    public abstract void load(final String namespace, final YamlConfiguration yaml);

    public abstract void addTranslations(
        final ResourcePackGenerator pack,
        final YamlConfiguration yaml,
        String langCode
    ) throws YamlLoadException;

    @SuppressWarnings("unchecked")
    public T get() {
        try {
            return (T) field.get(owner);
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Invalid field access on '" + field.getName() + "'. This is a bug.");
        }
    }

    @Override
    public String toString() {
        return (field.getDeclaringClass().getTypeName() + "::" + field.getName());
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/lang/LangField.kt
package org.oddlama.vane.core.lang

import org.bukkit.configuration.file.YamlConfiguration
import org.oddlama.vane.core.YamlLoadException
import org.oddlama.vane.core.module.Module
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator
import java.lang.reflect.Field

/**
 * Base abstraction for fields annotated as translatable language entries.
 *
 * @param T the runtime value type loaded into the target field.
 * @param module the module that owns this field.
 * @param owner the object instance that contains the reflected field.
 * @param field the reflected field to read/write.
 * @param mapName maps Java field names to YAML paths.
 */
abstract class LangField<T>(
    /** Owning module. */
    private val module: Module<*>,
    /** Reflective owner object containing the language field. */
    protected var owner: Any?,
    /** Reflected language field handle. */
    protected var field: Field,
    mapName: (String?) -> String
) {
    /**
     * YAML entry name used to load this language value.
     */
    var name: String
        protected set

    /**
     * Namespace used for translation keys.
     */
    private val namespace: String

    /**
     * Full translation key used in generated language files.
     */
    private val key: String

    init {
        if (!field.name.startsWith(PREFIX))
            throw RuntimeException(YamlLoadException.Lang("field must start with $PREFIX", this))
        this.name = mapName(field.name.substring(PREFIX.length))
        this.namespace = module.namespace()
        this.key = "$namespace.${yamlPath()}"
        field.isAccessible = true
    }

    /**
     * Returns the YAML path used to load this field.
     */
    fun yamlPath(): String = name

    /**
     * Validates that the YAML path exists.
     *
     * @param yaml the configuration to check.
     * @throws YamlLoadException if the path is missing.
     */
    @Throws(YamlLoadException::class)
    protected fun checkYamlPath(yaml: YamlConfiguration) {
        if (!yaml.contains(name, true))
            throw YamlLoadException.Lang("yaml is missing entry with path '$name'", this)
    }

    /**
     * Returns the owning module.
     */
    fun module(): Module<*> = module

    /**
     * Returns the translation namespace.
     */
    fun namespace(): String = namespace

    /**
     * Returns the fully qualified translation key.
     */
    fun key(): String = key

    /**
     * Validates that this field can be loaded from the YAML source.
     *
     * @param yaml the configuration to validate.
     * @throws YamlLoadException if validation fails.
     */
    @Throws(YamlLoadException::class)
    abstract fun checkLoadable(yaml: YamlConfiguration?)

    /**
     * Loads this field value from YAML into the owner object.
     *
     * @param namespace the active namespace.
     * @param yaml the configuration to read from.
     */
    abstract fun load(namespace: String?, yaml: YamlConfiguration?)

    /**
     * Adds translation values for this field into a resource pack builder.
     *
     * @param pack the target resource pack generator.
     * @param yaml the language configuration source.
     * @param langCode the target language code.
     * @throws YamlLoadException if translation data is invalid.
     */
    @Throws(YamlLoadException::class)
    abstract fun addTranslations(pack: ResourcePackGenerator?, yaml: YamlConfiguration?, langCode: String?)

    /**
     * Returns the current value from the underlying reflected field.
     */
    @Suppress("UNCHECKED_CAST")
    fun get(): T? =
        try {
            field.get(owner) as T?
        } catch (_: IllegalAccessException) {
            throw RuntimeException("Invalid field access on '${field.name}'. This is a bug.")
        }

    /**
     * Returns a stable debug representation of this reflected language field.
     */
    override fun toString(): String = "${field.declaringClass.typeName}::${field.name}"

    /**
     * Constants used by language field processing.
     */
    companion object {
        /**
         * Required field-name prefix for language entries.
         */
        const val PREFIX: String = "lang"
    }
}
